wp-includes/class-wp-service-workers.php

```diff
@@ -63,6 +129,117 @@ public function init() {
 		do_action_ref_array( 'wp_default_service_workers', array( &$this ) );
 	}
 
+	/**
+	 * Get workbox script.
+	 *
+	 * @return string Script.
+	 */
+	public function get_workbox_script() {
+
+		$workbox_dir = 'wp-includes/js/workbox-v3.4.1/';
+
+		$script = sprintf(
+			"importScripts( %s );\n",
+			wp_json_encode( PWA_PLUGIN_URL . $workbox_dir . 'workbox-sw.js', 64 /* JSON_UNESCAPED_SLASHES */ )
+		);
+
+		$options = array(
+			'debug'            => WP_DEBUG,
+			'modulePathPrefix' => PWA_PLUGIN_URL . $workbox_dir,
+		);
+		$script .= sprintf( "workbox.setConfig( %s );\n", wp_json_encode( $options, 64 /* JSON_UNESCAPED_SLASHES */ ) );
+
+		/**
+		 * Filters whether navigation preload is enabled.
+		 *
+		 * The filtered value will be sent as the Service-Worker-Navigation-Preload header value if a truthy string.
+		 * This filter should be set to return false to disable navigation preload such as when a site is using
+		 * the app shell model.
+		 *
+		 * @param bool|string $navigation_preload Whether to use navigation preload.
+		 */
+		$navigation_preload = apply_filters( 'service_worker_navigation_preload', true ); // @todo This needs to vary between admin and backend.
+		if ( false !== $navigation_preload ) {
+			if ( is_string( $navigation_preload ) ) {
+				$script .= sprintf( "workbox.navigationPreload.enable( %s );\n", wp_json_encode( $navigation_preload ) );
+			} else {
+				$script .= "workbox.navigationPreload.enable();\n";
+			}
+		} else {
+			$script .= "/* Navigation preload disabled. */\n";
+		}
+
+		$script .= '
+if ( ! self.wp ) {
+	self.wp = {};
+}
+
+wp.serviceWorker = workbox;';
+
+		return $script;
+	}
+
+	/**
+	 * Add precaching for wp-admin and wp-includes .js and .css files.
+	 */
+	protected function precache_admin_assets() {
+
+		$admin_dir   = ABSPATH . 'wp-admin/';
+		$admin_files = array_merge(
+			list_files( $admin_dir . 'css/' ),
+			list_files( $admin_dir . 'js/' ),
+			list_files( $admin_dir . 'images/' )
+		);
+		$inc_files   = array_merge(
+			list_files( ABSPATH . WPINC . '/js/' ),
+			list_files( ABSPATH . WPINC . '/css/' ),
+			list_files( ABSPATH . WPINC . '/images/' )
+		);
+
+		$routes = array_merge(
+			$this->get_routes_from_file_list( $admin_files, 'wp-admin' ),
+			$this->get_routes_from_file_list( $inc_files, 'wp-includes' )
+		);
+
+		if ( empty( $routes ) ) {
+			return;
+		}
+
+		$this->register_cached_route( $routes, self::STRATEGY_PRECACHE );
```

Comment:
It's going to be slow to as it is right now because it is going to fetch the file contents for all of the URLs to compute the hashes of each. Instead, we should use the current WordPress version as the revision, and get the list of files by looking at what is registered among default scripts and styles.

Reply:
Yes. It's what WordPress uses as the default `ver` as well: 

https://github.com/WordPress/wordpress-develop/blob/c52e7c79b2fd6381ba89fb8c40624e6621284743/src/wp-includes/script-loader.php#L64